**Hello everyone!**

This assignment is based on basics of python and then gradually moves on to the python libraries required for working in ML.

**Workflow**

Above each cell there are some instructions and resources (wherever necessary) for completing the question. You are supposed to solve the questions in the cells below the question text, feel free to add cells if you want.

Note: All the documentations provided are very extensive and you do not need to know everything but you are expected to be able to refer to them as and when required.


# Basic Python

If you are absolutely new to python, quickly go through this video: https://www.youtube.com/watch?v=K5KVEU3aaeQ&ab_channel=ProgrammingwithMosh

This will help you brush up your python to get your started.


# Part A



**Q.1**

 Create an array of 10 zeros.
Create an array of 10 fives.
 Create a 3x3 matrix with values ranging from 0 to 8
Use NumPy to generate a random number between 0 and 1
Create an array of 20 linearly spaced points between 0 and 1
  Write a program that counts how many vowels are in a string.





1.   Create an array of 10 zeros.
2.   Create an array of 10 fives.
3.   Create a 3x3 matrix with values ranging from 0 to 8
4.   Use NumPy to generate a random number between 0 and 1
5.   Create an array of 20 linearly spaced points between 0 and 1
7.   Write a program that counts how many vowels are in a string.





In [ ]:
## your code(for different question use different cells)
##like Q.1.
#1
import numpy as np
np.zeros(10)


array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [ ]:
#Q.1.2im

import numpy
numpy.ones(10)*5

array([5., 5., 5., 5., 5., 5., 5., 5., 5., 5.])

In [ ]:
import numpy
numpy.arange(9).reshape(3,3)

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [ ]:
import numpy
print(numpy.random.random)

<bound method RandomState.random of RandomState(MT19937) at 0x7F35FF00B640>


In [ ]:
import numpy
numpy.linspace(0,1,20)

array([0.        , 0.05263158, 0.10526316, 0.15789474, 0.21052632,
       0.26315789, 0.31578947, 0.36842105, 0.42105263, 0.47368421,
       0.52631579, 0.57894737, 0.63157895, 0.68421053, 0.73684211,
       0.78947368, 0.84210526, 0.89473684, 0.94736842, 1.        ])

In [ ]:
a=input("enter word")
vcount=0
for i in a:
  if i.lower() in "aeiou":
    vcount+=1
print(vcount)

enter wordhello
2


**Q.2**


1.   Extract the even numbers from this array using slicing.

numbers = [12, 7, 44, 19, 102, -8, 33, 56, 77, 90, 121, -24, 48]

2. Create a 2D array with 5 rows and 4 columns, filled with random integers between 10 and 100.
3. Extract a 3x2 subarray from the 2D array (e.g., first three rows, first two columns).
4. Reshape the 1D array of integers into a 4x5 2D array.
5. Flatten the 2D array back to a 1D array and print it.
6. Perform matrix multiplication between two 2D arrays (you can generate the second 2D array randomly).



In [ ]:
##your code
numbers = [12, 7, 44, 19, 102, -8, 33, 56, 77, 90, 121, -24, 48]
even=numbers[1::2]
print(even)

[7, 19, -8, 56, 90, -24]


In [ ]:
#2.2 and 2.3

import numpy as np
n=np.random.randint(10,100,size=(20)).reshape(5,4)
print(n)
print(n[:3,:2])

[[67 82 16 29]
 [94 25 59 78]
 [82 69 99 23]
 [57 14 51 62]
 [46 47 75 61]]
[[67 82]
 [94 25]
 [82 69]]


In [ ]:
import numpy as np
n=np.random.randint(10,100,size=(20)).reshape(4,5)
print(n)
m=n.flatten()
print(m)
k=np.random.randint(10,100,size=(20)).reshape(5,4)
print(k)
print(np.dot(n,k))

[[81 32 89 29 45]
 [20 46 38 39 16]
 [32 88 38 98 49]
 [23 42 97 96 37]]
[81 32 89 29 45 20 46 38 39 16 32 88 38 98 49 23 42 97 96 37]
[[83 87 76 68]
 [17 67 14 56]
 [55 23 76 18]
 [92 63 17 51]
 [11 39 44 17]]
[[15325 14820 15841 11146]
 [ 8296  8777  6419  6881]
 [15797 17639 10374 13619]
 [17197 14537 12968 11187]]


**Q.3**

(i) Check if matrix multiplication is valid

In [ ]:
#your code here
import numpy as np
A = np.random.randint(1, 100, size=(3, 5))
B = np.random.randint(1, 100, size=(5, 3))

if A.shape[1]==B.shape[0]:
  print("Matrix multiplication is valid")
  C = np.dot(A, B)
  print("Result:\n", C)
else:
    print("Matrix multiplication is NOT valid")


(ii) Using loops (without using numpy)

In [ ]:
#your code here
def matrix_multiply(A,B):
  array1 = np.array(A)
  array2 = np.array(B)
  box = []
  for i in range(0,array1.shape[0]):
    for j in range(0,array2.shape[1]):
      y=0
      for k in range(0,array1.shape[1]):
        y += array1[i][k] * array2[k][j]
        box.append(y)

**Q.4**
**Linear Regression:-** https://tahera-firdose.medium.com/linear-regression-derivation-of-slope-and-intercept-using-ordinary-least-square-971534ec6b77

(i) Given:

In [ ]:
X = [1,2,3,4,5]
y = [2,4,5,4,5]

sum_x = 0
sum_y = 0
count_x=0
count_y=0
for i in X:
  sum_x+=i
  count_x+=1
for j in y:
  sum_y+=j
  count_y+=1
mean_x = sum_x/count_x
mean_y = sum_y/count_y
print(mean_x)
print(mean_y)
summ=0
sumn=0
for i in range(len(X)):
  summ+=(X[i]-mean_x)*(y[i]-mean_y)
  sumn+=(X[i]-mean_x)**2
m=summ/sumn
c=mean_y-m*mean_x
print(m)
print(c)

3.0
4.0
0.6
2.2


Compute:

mean of X

mean of y

slope (m)

intercept (c)

predicted values

No libraries allowed.(sklearn or numpy)


##  Pandas


csv file:- https://drive.google.com/file/d/1Q4bT1VNcZFY-tA3x3BV8MwqezuhqVFLq/view?usp=sharing


**Perform the following operations on the csv file:**

1. Read the csv into a pandas dataframe and back it up (you will thank yourself for doing this when you mess up something while manipulating the dataframe).
2. Explore the dataframe (find out the different ways you can do this).
3. Play around with each column of the dataframe and figure out ways to modify them according to your needs.
4. Read up on ``iloc`` and demonstrate it on the dataframe.
5. Handle ``null`` values in the dataframe.
6. Try dropping rows/columns.
7. Filter rows where a string column starts with a specific letter.
8. Create a new column as a combination of two existing columns (this is feature engineering which you may use in your ML journey).

# Task
Define two matrices, check if matrix multiplication is valid by comparing their dimensions, and then implement matrix multiplication for these matrices using nested loops (without NumPy). Finally, present the results.

## Check Matrix Multiplication Validity

### Subtask:
Define two matrices and write code to determine if matrix multiplication is valid by comparing their dimensions (number of columns in the first matrix must equal the number of rows in the second matrix).


**Reasoning**:
The subtask requires defining two matrices and checking if their dimensions allow for matrix multiplication. This code block will define the matrices, extract their dimensions, and perform the validity check as instructed.



In [ ]:
matrix1 = [[1, 2, 3], [4, 5, 6]] # 2x3 matrix
matrix2 = [[7, 8], [9, 10], [11, 12]] # 3x2 matrix

rows1 = len(matrix1)
cols1 = len(matrix1[0])

rows2 = len(matrix2)
cols2 = len(matrix2[0])

print(f"Matrix 1 dimensions: {rows1}x{cols1}")
print(f"Matrix 2 dimensions: {rows2}x{cols2}")

if cols1 == rows2:
    print("Matrix multiplication is valid.")
else:
    print("Matrix multiplication is not valid.")

# Example of incompatible matrices
print("\n--- Testing incompatible matrices ---")
matrix3 = [[1, 2], [3, 4]] # 2x2 matrix
matrix4 = [[5, 6, 7], [8, 9, 10]] # 2x3 matrix (compatible, just for example)
matrix5 = [[1,2,3],[4,5,6]] # 2x3 matrix
matrix6 = [[7,8],[9,10]] # 2x2 matrix

rows5 = len(matrix5)
cols5 = len(matrix5[0])

rows6 = len(matrix6)
cols6 = len(matrix6[0])

print(f"Matrix 5 dimensions: {rows5}x{cols5}")
print(f"Matrix 6 dimensions: {rows6}x{cols6}")

if cols5 == rows6:
    print("Matrix multiplication is valid.")
else:
    print("Matrix multiplication is not valid.")

Matrix 1 dimensions: 2x3
Matrix 2 dimensions: 3x2
Matrix multiplication is valid.

--- Testing incompatible matrices ---
Matrix 5 dimensions: 2x3
Matrix 6 dimensions: 2x2
Matrix multiplication is not valid.


## Matrix Multiplication using Loops

### Subtask:
Implement matrix multiplication for the defined matrices using nested loops, without relying on the NumPy library. This involves iterating through rows, columns, and the inner product dimension.


**Reasoning**:
The subtask requires implementing matrix multiplication using nested loops. This code block will initialize the result matrix and then perform the multiplication as specified in the instructions, without using NumPy.



In [ ]:
if cols1 == rows2:
    # Initialize result matrix with zeros
    result_matrix = [[0 for _ in range(cols2)] for _ in range(rows1)]

    # Perform matrix multiplication
    for i in range(rows1):
        for j in range(cols2):
            for k in range(cols1):
                result_matrix[i][j] += matrix1[i][k] * matrix2[k][j]

    print("Result of matrix multiplication:")
    for row in result_matrix:
        print(row)
else:
    print("Matrix multiplication is not valid for the given matrices.")

Result of matrix multiplication:
[58, 64]
[139, 154]


## Final Task

### Subtask:
Confirm that both parts of Q3 have been addressed and present the results of the matrix multiplication.


## Summary:

### Q&A
1.  **Is matrix multiplication valid for the given matrices?**
    For `matrix1` (2x3) and `matrix2` (3x2), matrix multiplication is valid because the number of columns in `matrix1` (3) equals the number of rows in `matrix2` (3).
2.  **What is the result of the matrix multiplication?**
    The result of multiplying `matrix1` by `matrix2` is `[[58, 64], [139, 154]]`.

### Data Analysis Key Findings
*   The initial `matrix1` (2x3) and `matrix2` (3x2) were confirmed as valid for multiplication, as the number of columns in `matrix1` (3) matched the number of rows in `matrix2` (3).
*   An example of incompatible matrices, `matrix5` (2x3) and `matrix6` (2x2), correctly determined that multiplication is not valid due to a mismatch in dimensions (3 != 2).
*   The matrix multiplication for `matrix1` and `matrix2` was successfully performed using a triple nested loop, resulting in a 2x2 matrix: `[[58, 64], [139, 154]]`.

### Insights or Next Steps
*   This exercise effectively demonstrates the foundational logic behind matrix multiplication without relying on specialized libraries, reinforcing core programming and mathematical concepts.
*   For real-world applications involving linear algebra and large-scale matrix operations, transitioning to optimized libraries such as NumPy is recommended for significantly improved computational efficiency and reduced development time.
